### PR TITLE
Duet AI Transform service SDK Harness prompts

### PR DESCRIPTION
**Pull request by olehborysevych**
Transform service lookup prompts

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+```java
+--transformsToOverride=beam:transform:org.apache.beam:bigquery_read:v1 --transformServiceBeamVersion=2.xy.z
+```
+
+In the above example, `--transformsToOverride` specifies the URN of the transform you want to upgrade or downgrade, and `--transformServiceBeamVersion` specifies the Beam version you want to upgrade to.
+
+The framework will automatically download the specified version of Docker containers for the transforms and use them in the pipeline. You must have Docker installed on the machine that starts the service.
+
+Currently the following transforms are supported:
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:17]:
```suggestion
Currently, the Transform service can upgrade the following transforms:
```

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:18]:
```suggestion
```

File: learning/prompts/documentation-lookup-nolinks/48_sdk_harness.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+Apache Beam allows configuration of the SDK harness to accommodate varying cluster setups:
+
+* **environment_type**: determines where user code will be executed:
+  * **DOCKER**: User code is executed within a container started on each worker node. This requires docker to be installed on worker nodes (default). Use `environment_config`  to specify the Docker image URL. Official Docker images are used by default. Alternatively, you can build your own image. Prebuilt SDK container images are released per supported language during Beam releases and pushed to Docker Hub.
+  * **PROCESS**: User code is executed by processes that are automatically started by the runner on each worker node.
+  * **EXTERNAL**: User code will be dispatched to an external service. Use `environment_config` to specify the address for the external service, e.g. `localhost:50000`.
+  * **LOOPBACK**: User code is executed within the same process that submitted the pipeline.
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/48_sdk_harness.md:18]:
```suggestion
```

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+* BigQuery read transform: `beam:transform:org.apache.beam:bigquery_read:v1`
+* BigQuery write transform: `beam:transform:org.apache.beam:bigquery_write:v1`
+* Kafka read transform: `beam:transform:org.apache.beam:kafka_read_with_metadata:v2`
+* Kafka write transform: `beam:transform:org.apache.beam:kafka_write:v2`
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:22]:
```suggestion
* BigQuery read: `beam:transform:org.apache.beam:bigquery_read:v1`
* BigQuery write: `beam:transform:org.apache.beam:bigquery_write:v1`
* Kafka read: `beam:transform:org.apache.beam:kafka_read_with_metadata:v2`
* Kafka write: `beam:transform:org.apache.beam:kafka_write:v2`
```

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+Transform service implement Beam expansion API. This means you can use the Transform service to construct and execute multi-language pipelines. For example, you can build a Python pipeline that uses Java `KafkaIO` transform and execute in without installing Java locally.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:24]:
```suggestion
The Transform service implements the Beam expansion API, enabling multi-language pipelines to leverage it for expanding supported transforms. This feature allows you to create and run multi-language pipelines without additional language runtimes. For instance, you can build a Python pipeline that utilizes a Java `KafkaIO` transform without the need to install Java locally.
```

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+Transform service can be started automatically by Apache Beam SDK or manually by users:
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:26]:
```suggestion
In some cases, Apache Beam SDKs can start the Transform service automatically, provided that Docker is available locally. You can also start the Transform service manually by running the following command:
```

File: learning/prompts/documentation-lookup-nolinks/49_transform_service.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup-nolinks/49_transform_service.md:57]:
```suggestion
The Beam Transform service includes several portable transforms implemented in the Apache Beam Java and Python SDKs:
* Java transforms: Google Cloud I/O connectors, the Kafka I/O connector, and the JDBC I/O connector.
* Python transforms: all portable transforms implemented within the Apache Beam Python SDK, such as RunInference and DataFrame transforms.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+Prompt:
+What is SDK Harness in Apache Beam?
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:2]:
```suggestion
What is SDK harness in Apache Beam?
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+Response:
+Interoperability between Apache Beam SDKs and runners is a key aspect of Apache Beam. The [portability framework](https://beam.apache.org/roadmap/portability/) introduces well-defined, language-neutral data structures and protocols between the SDK and runner. This interop layer – called the `Portability API` – ensures that SDKs and runners can work with each other uniformly, reducing the interoperability burden for both SDKs and runners to a constant effort.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:5]:
```suggestion
Interoperability between Apache Beam SDKs and runners is a key aspect of Apache Beam. The [portability framework](https://beam.apache.org/roadmap/portability/) introduces well-defined, language-neutral data structures and protocols between the SDK and runner. This interoperability layer, known as the 'portability API', ensures that SDKs and runners can seamlessly work with each other, reducing the interoperability burden for both SDKs and runners to a constant effort.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+The portability API consists of a set of smaller contracts that isolate SDKs and runners for job submission, management and execution. These contracts use `protobuf`s and `gRPC` for broad language support.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:7]:
```suggestion
The portability API consists of a set of smaller contracts that isolate SDKs and runners for job submission, management, and execution. These contracts utilize protocols like `protobuf` and `gRPC` to provide broad language support.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+All SDKs currently support the portability framework. See the [Portability support table](https://docs.google.com/spreadsheets/d/1KDa_FGn1ShjomGd-UUDOhuh2q73de2tPz6BqHpzqvNI/edit#gid=0) for details.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:9]:
```suggestion
Currently, all SDKs support the portability framework. For the latest information on portability support across SDKs, features, and runners, refer to the [Apache Beam Portability Support Matrix](https://docs.google.com/spreadsheets/d/1KDa_FGn1ShjomGd-UUDOhuh2q73de2tPz6BqHpzqvNI/edit#gid=0).
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+The SDK harness is a SDK-provided program responsible for executing user code and is run separately from the runner. SDK harness initialization relies on the Provision and `Artifact API`s for obtaining staged files, pipeline options and environment information.
+
+Apache Beam allows configuration of the SDK harness to accommodate varying cluster setups:
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:13]:
```suggestion
Apache Beam offers configuration options for the SDK harness to cater to diverse cluster setups. These options include:
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:14]:
```suggestion
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+* **environment_type**: determines where user code will be executed:
+  * **DOCKER**: User code is executed within a container started on each worker node. This requires docker to be installed on worker nodes (default). Use `environment_config`  to specify the Docker image URL. Official Docker images are used by default. Alternatively, you can build your own image by following the instructions [here](https://beam.apache.org/documentation/runtime/environments/). Prebuilt SDK container images are released per supported language during Beam releases and pushed to [Docker Hub](https://hub.docker.com/search?q=apache%2Fbeam&type=image).
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:16]:
```suggestion
  * `DOCKER`: executes user code within a container on each worker node. Docker must be installed on worker nodes. You can specify the Docker image URL using the `environment_config` parameter. Prebuilt SDK container images are available with each Apache Beam release and pushed to [Docker Hub](https://hub.docker.com/search?q=apache%2Fbeam&type=image). You can also [build your custom image](https://beam.apache.org/documentation/runtime/environments/).
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  * **PROCESS**: User code is executed by processes that are automatically started by the runner on each worker node.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:17]:
```suggestion
  * `PROCESS`: executes user code through processes that are automatically initiated by the runner on each worker node.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  * **EXTERNAL**: User code will be dispatched to an external service. Use `environment_config` to specify the address for the external service, e.g. `localhost:50000`.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:18]:
```suggestion
  * `EXTERNAL`: dispatches user code to an external service. Use the `environment_config` parameter to specify the service address, for example, `localhost:50000`.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  * **LOOPBACK**: User code is executed within the same process that submitted the pipeline.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:19]:
```suggestion
  * `LOOPBACK`: executes user code within the same process that submitted the pipeline.
```

File: learning/prompts/documentation-lookup/48_sdk_harness.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+* **sdk_worker_parallelism**:  sets the number of SDK workers that run on each worker node. The default is 1. If 0, the value is automatically set by the runner by looking at different parameters, such as the number of CPU cores on the worker machine.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/48_sdk_harness.md:21]:
```suggestion
2. **`sdk_worker_parallelism`**: determines the number of SDK workers per worker node. The default value is 1, but setting it to 0 enables automatic determination by the runner based on factors like the number of CPU cores on the worker machine.
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
+Prompt:
+What is a Transformer Service in Apache Beam?
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:2]:
```suggestion
What is Transform service in Apache Beam?
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+Response:
+The `Transform service` is [Docker Compose service](https://docs.docker.com/compose/) included into Apache Beam SDK versions 2.49.0 and later. You can use Transform service to upgrade or downgrade Beam SDK version of individual supported transforms in your pipeline without changing the Beam version of the pipeline. This feature is only supported for Beam Java SDK 2.53.0 and later.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:5]:
```suggestion
The Transform service is a [Docker Compose](https://docs.docker.com/compose/) service included in Apache Beam SDK versions 2.49.0 and later. It enables you to upgrade or downgrade the Beam SDK version of individual supported transforms in your pipeline without changing the overall Beam version of the pipeline. Additionally, you can utilize the Transform service to create and execute multi-language pipelines without needing to install support for additional language runtimes.
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+To use this feature simply execute Java pipeline with additional pipeline options specifying the URNs of the transforms you want to upgrade and the Beam version you want to upgrade to:
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:7]:
```suggestion
You can utilize the Transform service to upgrade specific transforms only if you are using Beam Java SDK 2.53.0 and later. To employ this feature, execute a Java pipeline with additional pipeline options specifying the URNs of the transforms you want to upgrade and the desired Apache Beam version:
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+```java
+--transformsToOverride=beam:transform:org.apache.beam:bigquery_read:v1 --transformServiceBeamVersion=2.xy.z
+```
+
+In the above example, `--transformsToOverride` specifies the URN of the transform you want to upgrade or downgrade, and `--transformServiceBeamVersion` specifies the Beam version you want to upgrade to.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:13]:
```suggestion
In the provided example, `--transformsToOverride` specifies the URN of the transform to upgrade or downgrade, while `--transformServiceBeamVersion` specifies the target Beam version.
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+The framework will automatically download the specified version of Docker containers for the transforms and use them in the pipeline. You must have Docker installed on the machine that starts the service.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:15]:
```suggestion
The framework automatically downloads the specified version of Docker containers for the transforms and uses them in the pipeline. You must have Docker installed on the machine that starts the service.
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+Currently the following transforms are supported:
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:18]:
```suggestion
Currently, the Transform service can upgrade the following transforms:
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+* BigQuery read transform: `beam:transform:org.apache.beam:bigquery_read:v1`
+* BigQuery write transform: `beam:transform:org.apache.beam:bigquery_write:v1`
+* Kafka read transform: `beam:transform:org.apache.beam:kafka_read_with_metadata:v2`
+* Kafka write transform: `beam:transform:org.apache.beam:kafka_write:v2`
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:23]:
```suggestion
* BigQuery read: `beam:transform:org.apache.beam:bigquery_read:v1`
* BigQuery write: `beam:transform:org.apache.beam:bigquery_write:v1`
* Kafka read: `beam:transform:org.apache.beam:kafka_read_with_metadata:v2`
* Kafka write: `beam:transform:org.apache.beam:kafka_write:v2`

```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+Transform service implement Beam expansion API. This means you can use the Transform service to construct and execute multi-language pipelines. For example, you can build a Python pipeline that uses Java `KafkaIO` transform and execute in without installing Java locally.
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:24]:
```suggestion
The Transform service implements the Beam expansion API, enabling multi-language pipelines to leverage it for expanding supported transforms. This feature allows you to create and run multi-language pipelines without additional language runtimes. For instance, you can build a Python pipeline that utilizes a Java `KafkaIO` transform without the need to install Java locally.
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+Transform service can be started automatically by Apache Beam SDK or manually by users:
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:26]:
```suggestion
In some cases, Apache Beam SDKs can start the Transform service automatically, provided that Docker is available locally. You can also start the Transform service manually by running the following command:
```

File: learning/prompts/documentation-lookup/49_transform_service.md [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/49_transform_service.md:57]:
```suggestion
The Beam Transform service includes several portable transforms implemented in the Apache Beam Java and Python SDKs:
* Java transforms: Google Cloud I/O connectors, the Kafka I/O connector, and the JDBC I/O connector.
* Python transforms: all portable transforms implemented within the Apache Beam Python SDK, such as RunInference and DataFrame transforms.
```

**Comment by olehborysevych** [comment]
Hey @damccorm this one is also ready for review

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**